components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java

```diff
@@ -0,0 +1,203 @@
+import org.wso2.carbon.identity.oauth2.internal.OAuth2ServiceComponentHolder;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.oauth.common.OAuthConstants.OAuth20Params.USERINFO;
+
+public abstract class AbstractUserInfoResponseBuilder implements UserInfoResponseBuilder {
+
+    @Override
+    public final String getResponseString(OAuth2TokenValidationResponseDTO tokenResponse)
+            throws UserInfoEndpointException, OAuthSystemException {
+
+        String spTenantDomain = getServiceProviderTenantDomain(tokenResponse);
+        // Retrieve user claims.
+        Map<String, Object> userClaims = retrieveUserClaims(tokenResponse);
+
+        // Filter user claims based on the requested scopes
+        Map<String, Object> filteredUserClaims =
+                getUserClaimsFilteredByScope(tokenResponse.getScope(), spTenantDomain, userClaims);
+
+        // Handle subject claim.
+        String subjectClaim = getSubjectClaim(userClaims, spTenantDomain, tokenResponse);
+        filteredUserClaims.put(OAuth2Util.SUB, subjectClaim);
+
+        // Handle essential claims
+        Map<String, Object> essentialClaims = getEssentialClaims(tokenResponse, userClaims);
+        filteredUserClaims.putAll(essentialClaims);
+
+        return buildResponse(tokenResponse, spTenantDomain, filteredUserClaims);
+    }
+
+
+    protected String getSubjectClaim(Map<String, Object> userClaims,
+                                     String spTenantDomain,
+                                     OAuth2TokenValidationResponseDTO tokenResponse) throws UserInfoEndpointException {
+
+        String subjectClaim = (String) userClaims.get(OAuth2Util.SUB);
+        if (StringUtils.isBlank(subjectClaim)) {
+            // We need to send the tenant aware and user store domain removed name.
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(tokenResponse.getAuthorizedUser());
+            subjectClaim = UserCoreUtil.removeDomainFromName(tenantAwareUsername);
+        }
+        return returnSubjectClaim(subjectClaim, spTenantDomain, tokenResponse);
+    }
+
+    private String returnSubjectClaim(String sub,
+                                      String tenantDomain,
+                                      OAuth2TokenValidationResponseDTO tokenResponse) throws UserInfoEndpointException {
+
+        String clientId = getClientId(tokenResponse);
+        ServiceProvider serviceProvider = getServiceProvider(tenantDomain, clientId);
+
+        String userTenantDomain = MultitenantUtils.getTenantDomain(tokenResponse.getAuthorizedUser());
+        String userStoreDomain = UserCoreUtil.extractDomainFromName(tokenResponse.getAuthorizedUser());
```

Comment:
Use IdentityUtil.extractDomainFromName instead of UserCoreUtil to avoid the hard corded PRIMARY domain